csrc/multidevice/utils.cpp

```diff
@@ -151,6 +217,116 @@ void insertReshardings(Fusion* fusion) {
   }
 }
 
+void insertShardedAxisReordering(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  std::vector<Expr*> reshard_exprs;
+  for (auto expr : exprs) {
+    if (isResharding(expr)) {
+      reshard_exprs.push_back(expr);
+    }
+  }
+  for (auto expr : reshard_exprs) {
+    NVF_ERROR(
+        ir_utils::isTvOp(expr),
+        "Non-tv op is not supported yet: ",
+        expr->toString());
+    NVF_ERROR(
+        expr->outputs().size() == 1,
+        "Resharding operations can only have one output");
+    NVF_ERROR(
+        expr->inputs().size() == 1,
+        "Resharding operations can have only one input");
+    auto output = expr->outputs().at(0)->as<TensorView>();
+    auto input = expr->inputs().at(0)->as<TensorView>();
+    auto [shard_additions, shard_deletions] = getShardingChanges(expr);
+    NVF_ERROR(
+        shard_additions.size() + shard_deletions.size() <= 1,
+        "Resharding expr can only support one axis")
+
+    // For gather operations i.e. ID goes from sharded to unsharded
+    // this will rematerialize a sharded axis.
+    // ProcessGroup expects contiguous tensors.
+    // Update input to push the rematerialized axis to the front -> collective
+    // -> permute the rematerizlied axis to the proper location Example: [i0
+    // DIDx(i1)] -> [i0 i1] Rewritten to: [DIDx(i1) i0] -> [i1 i0] -> [i0 i1]
+    // Note: there are no reduction based collectives that materializes an axis
+    // so expr is guaranteed to be a set.
+    if (!shard_deletions.empty()) {
+      IterDomain* shard_deleted_id = shard_deletions[0];
+      int sharding_axis =
+          static_cast<int>(input->domain()->rootPosOf(shard_deleted_id));
+      if (isOutermostAllocatedId(input, shard_deleted_id)) {
+        continue;
+      }
+      TensorView* input_permute = permute(input, {{sharding_axis, 0}});
+      TensorView* output_permute = set(input_permute);
+      TensorView* new_output = permute(output_permute, {{0, sharding_axis}});
+      ir_utils::replaceValInAllExprInputsAndFusionOutputs(output, new_output);
```

Comment:
I am afraid there is a problem here if a tensor is the input of two resharding expr. In this case the second expr will be replaced by a new one after this line, and so the for loop will not iterate over it. Am I correct?

Same below for the other branch

Reply:
The expression that take `output` as input will not be replaced, instead `new_output` will be the input tv instead of `output`. `new_output` is also identical to `output` just generated by a different expression.